nexus/reconfigurator/planning/src/blueprint_builder/builder.rs

```diff
@@ -3721,4 +3811,352 @@ pub mod test {
+
+        // Add another Nexus zone with same image source - should reuse generation
+        builder
+            .sled_add_zone_nexus_with_config(
+                second_sled_id,
+                false,
+                vec![],
+                image_source.clone(),
+                builder.parent_blueprint().nexus_generation,
+            )
+            .expect("failed to add nexus zone");
+
+        let blueprint2 = builder.build();
+        verify_blueprint(&blueprint2);
+
+        // Find all nexus zones and verify they have the same generation
+        let nexus_zones: Vec<_> = blueprint2
+            .all_omicron_zones(BlueprintZoneDisposition::any)
+            .filter_map(|(_, zone)| match &zone.zone_type {
+                BlueprintZoneType::Nexus(nexus) => Some(nexus),
+                _ => None,
+            })
+            .collect();
+
+        assert_eq!(nexus_zones.len(), 2);
+        assert_eq!(nexus_zones[0].nexus_generation, existing_nexus_gen);
+        assert_eq!(nexus_zones[1].nexus_generation, existing_nexus_gen);
+
+        logctx.cleanup_successful();
+    }
+
+    /// Test nexus generation assignment logic for different image sources
+    #[test]
+    fn test_nexus_generation_assignment_different_image_increment() {
+        static TEST_NAME: &str =
+            "test_nexus_generation_assignment_different_image_increment";
+        let logctx = test_setup_log(TEST_NAME);
+        let mut rng = SimRngState::from_seed(TEST_NAME);
+
+        // Start with a system that has one Nexus zone
+        let (example_system, blueprint) =
+            ExampleSystemBuilder::new(&logctx.log, TEST_NAME)
+                .nexus_count(1)
+                .build();
+        verify_blueprint(&blueprint);
+
+        // Get the generation of the existing nexus zone
+        let existing_nexus_gen = blueprint
+            .all_omicron_zones(BlueprintZoneDisposition::any)
+            .find_map(|(_, zone)| match &zone.zone_type {
+                BlueprintZoneType::Nexus(nexus) => {
+                    assert_eq!(
+                        zone.image_source,
+                        BlueprintZoneImageSource::InstallDataset
+                    );
+                    Some(nexus.nexus_generation)
+                }
+                _ => None,
+            })
+            .expect("should have found existing nexus");
+
+        let mut builder = BlueprintBuilder::new_based_on(
+            &logctx.log,
+            &blueprint,
+            &example_system.input,
+            &example_system.collection,
+            "test",
+            rng.next_planner_rng(),
+        )
+        .expect("failed to create builder");
+
+        // Get a different sled
+        let sled_ids: Vec<_> = example_system
+            .input
+            .all_sled_ids(SledFilter::Commissioned)
+            .collect();
+        let second_sled_id = sled_ids[1];
+
+        // Use a different image source (artifact vs install dataset)
```

Comment:
Actually, I wonder if this should produce an error if any Nexus zones have source "install dataset"?  Because we don't know that the image is actually different.